File: api/internal/platform/r2.go
package platform

import (
	"bytes"
	"context"
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/miguel-anay/career-ops-saas/api/internal/config"
)

// R2Client is an S3-compatible client for Cloudflare R2.
type R2Client struct {
	client *s3.Client
	bucket string
}

// NewR2Client creates a new R2Client using the provided config.
// Returns an error if R2 credentials are missing or client cannot be created.
func NewR2Client(cfg *config.Config) (*R2Client, error) {
	if cfg.R2AccountID == "" || cfg.R2AccessKeyID == "" || cfg.R2SecretAccessKey == "" || cfg.R2Bucket == "" {
		return nil, fmt.Errorf("R2 configuration incomplete: R2_ACCOUNT_ID, R2_ACCESS_KEY_ID, R2_SECRET_ACCESS_KEY, R2_BUCKET are required")
	}

	endpoint := fmt.Sprintf("https://%s.r2.cloudflarestorage.com", cfg.R2AccountID)

	r2Resolver := aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
		return aws.Endpoint{
			URL: endpoint,
		}, nil
	})

	awsCfg, err := awsconfig.LoadDefaultConfig(context.Background(),
		awsconfig.WithEndpointResolverWithOptions(r2Resolver),
		awsconfig.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
			cfg.R2AccessKeyID,
			cfg.R2SecretAccessKey,
			"",
		)),
		awsconfig.WithRegion("auto"),
	)
	if err != nil {
		return nil, fmt.Errorf("load aws config for R2: %w", err)
	}

	client := s3.NewFromConfig(awsCfg, func(o *s3.Options) {
		o.UsePathStyle = true
	})

	return &R2Client{
		client: client,
		bucket: cfg.R2Bucket,
	}, nil
}

// SignedDownloadURL generates a pre-signed URL for downloading an object from R2.
func (r *R2Client) SignedDownloadURL(key string, expiry time.Duration) (string, error) {
	presignClient := s3.NewPresignClient(r.client)

	req, err := presignClient.PresignGetObject(context.Background(), &s3.GetObjectInput{
		Bucket: aws.String(r.bucket),
		Key:    aws.String(key),
	}, s3.WithPresignExpires(expiry))
	if err != nil {
		return "", fmt.Errorf("presign get object: %w", err)
	}

	return req.URL, nil
}

// UploadObject uploads a byte slice to R2 at the given key with the specified content type.
func (r *R2Client) UploadObject(ctx context.Context, key string, body []byte, contentType string) error {
	_, err := r.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:      aws.String(r.bucket),
		Key:         aws.String(key),
		Body:        bytes.NewReader(body),
		ContentType: aws.String(contentType),
	})
	if err != nil {
		return fmt.Errorf("upload object to R2: %w", err)
	}
	return nil
}

File: worker/lib/r2.mjs
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3'
import 'dotenv/config'

const s3 = new S3Client({
  region: 'auto',
  endpoint: process.env.R2_ENDPOINT,
  credentials: {
    accessKeyId: process.env.R2_ACCESS_KEY_ID,
    secretAccessKey: process.env.R2_SECRET_ACCESS_KEY,
  },
})

const BUCKET = process.env.R2_BUCKET

/**
 * Upload a buffer to Cloudflare R2 (S3-compatible).
 *
 * @param {string} key - Object key (path within the bucket)
 * @param {Buffer} buffer - File content as a Buffer
 * @param {string} mimeType - MIME type (e.g., 'application/pdf')
 * @returns {Promise<{key: string, etag: string}>}
 */
export async function uploadBuffer(key, buffer, mimeType) {
  const command = new PutObjectCommand({
    Bucket: BUCKET,
    Key: key,
    Body: buffer,
    ContentType: mimeType,
  })

  const result = await s3.send(command)

  return {
    key,
    etag: result.ETag || '',
  }
}
